Failure "no tls auth payload in config"

openvpn-client-lwt client1.cfg 
openvpn-client-lwt: [WARNING] IGNORING "status" "openvpn-status.log"
openvpn-client-lwt: [WARNING] IGNORING "dh" "dh2048.pem"
openvpn-client-lwt: [WARNING] IGNORING "status" "openvpn-status.log"
openvpn-client-lwt: [WARNING] IGNORING "dh" "dh2048.pem"
openvpn-client-lwt: [WARNING] keepalive 10 120 overrides former ping interval 10
openvpn-client-lwt: [ERROR] client construction failed no tls auth payload in config
openvpn_client: internal error, uncaught exception:
                (Failure "no tls auth payload in config")
                Raised at file "src/core/lwt.ml", line 3027, characters 20-29
                Called from file "src/unix/lwt_main.ml", line 27, characters 10-20
                Called from file "src/unix/lwt_main.ml", line 114, characters 8-13
                Re-raised at file "src/unix/lwt_main.ml", line 120, characters 4-13
                Called from file "cmdliner_term.ml", line 25, characters 19-24
                Called from file "cmdliner.ml", line 33, characters 27-34
                Called from file "cmdliner.ml", line 117, characters 32-39

config:
auth-user-pass userpass.txt
tls-client
dev tun
persist-key
persist-tun
topology subnet
#port 1194
remote localhost 1195
proto udp
keepalive 10 120
ca ca.public.certificate
cert client1.public.certificate
key client1.secret.key
#user ovpn
#group ovpn
#auth SHA512
tls-version-min 1.2
#tls-cipher TLS-DHE-RSA-WITH-AES-256-GCM-SHA384:TLS-DHE-RSA-WITH-AES-256-CBC-SHA256
#ncp-ciphers AES-256-GCM:AES-256-CBC
status openvpn-status.log
#log /var/log/openvpn-assym.log
verb 3
dh dh2048.pem

tls-client
dev tun
persist-key
persist-tun
topology subnet
#port 1194
remote localhost 1194
proto udp
keepalive 10 120
ca ca.public.certificate
cert client1.public.certificate
key client1.secret.key
#user ovpn
#group ovpn
#auth SHA512
tls-version-min 1.2
#tls-cipher TLS-DHE-RSA-WITH-AES-256-GCM-SHA384:TLS-DHE-RSA-WITH-AES-256-CBC-SHA256
#ncp-ciphers AES-256-GCM:AES-256-CBC
status openvpn-status.log
#log /var/log/openvpn-assym.log
verb 3
dh dh2048.pem
pull

--- a/src/config.ml
+++ b/src/config.ml
@@ -594,7 +594,10 @@ let inline_payload element =
       | Invalid_argument msg -> abort msg)
     (string "-----END OpenVPN Static key V1-----" *> a_newline
      <|>abort "Missing END mark")
-  <* (end_of_input <|> abort "Data after -----END mark")
+  <* commit <* (
+    (skip_many (a_newline <|> a_whitespace) *> end_of_input)
+    <|> (pos >>= fun i ->
+         abort (Fmt.strf "Data after -----END mark at byte offset %d" i)))
   >>= (fun lst ->
       let sz = Cstruct.lenv lst in
       if 256 = sz then return lst else
@@ -1152,13 +1155,21 @@ let parse_inline str = let open Rresult in
                    (string_of_inlineable kind))
 
 let eq : eq = { f = fun (type x) (k: x k) (v:x) (v2:x) ->
-    let eq = v = v2 in
-    (*TODO non-polymorphic comparison*)
-    begin if not eq then Logs.debug
+    begin match k, v, v2 with
+      | Secret, (a,b,c,d), (a',b',c',d') ->
+        Cstruct.equal a a' && Cstruct.equal b b'
+        && Cstruct.equal c c' && Cstruct.equal d d'
+      | Tls_auth, (dir, a,b,c,d), (dir', a',b',c',d') ->
+        dir = dir' && Cstruct.equal a a' && Cstruct.equal b b'
+        && Cstruct.equal c c' && Cstruct.equal d d'
+      | _ ->     (*TODO non-polymorphic comparison*)
+        let eq = (v = v2) in
+        Logs.debug
           (fun m -> m "eq self-test: @[<v>%a@,<>@,%a@]"
               pp (singleton k v)
-              pp (singleton k v2))
-        ; eq end }
+              pp (singleton k v2)) ;
+        eq
+    end }
 
 
 let resolve_conflict (type a) t (k:a key) (v:a)

--- a/test/config_tests.ml
+++ b/test/config_tests.ml
@@ -193,6 +193,43 @@ testpass
   let _ip3 = Ipaddr.of_string_exn "10.0.42.5" in
   () (* TODO check that the ports and remotes also match the written *)
 
+let whitespace_after_tls_auth () =
+  let expected = Openvpn.Config.add Tls_auth
+      (None,
+       Cstruct.create 64, Cstruct.create 64,
+       Cstruct.create 64, Cstruct.create 64) minimal_config in
+  let with_newlines =
+    Fmt.strf {|%a
+tls-auth [inline]
+<tls-auth>
+-----BEGIN OpenVPN Static key V1-----
+00000000000000000000000000000000
+00000000000000000000000000000000
+00000000000000000000000000000000
+00000000000000000000000000000000
+00000000000000000000000000000000
+00000000000000000000000000000000
+00000000000000000000000000000000
+00000000000000000000000000000000
+00000000000000000000000000000000
+00000000000000000000000000000000
+00000000000000000000000000000000
+00000000000000000000000000000000
+00000000000000000000000000000000
+00000000000000000000000000000000
+00000000000000000000000000000000
+00000000000000000000000000000000
+-----END OpenVPN Static key V1-----
+
+
+</tls-auth>
+|} Openvpn.Config.pp minimal_config
+  in
+  Alcotest.(check (result conf_map pmsg))
+    "Allow whitespace after ----END of tls-auth"
+    (Ok expected)
+    (parse_noextern with_newlines)
+
 let crowbar_fuzz_config () =
   Crowbar.add_test ~name:"Fuzzing doesn't crash Config.parse_client"
     [Crowbar.bytes] (fun s ->
@@ -206,5 +243,7 @@ let tests = [
   "auth-user-pass trailing whitespace", `Quick,
   auth_user_pass_trailing_whitespace ;
   "rport precedence", `Quick, rport_precedence ;
+  "trailing whitespace after <tls-auth>", `Quick,
+  whitespace_after_tls_auth;
   "crowbar fuzzing", `Slow, crowbar_fuzz_config ;
 ]

